### PR TITLE
array_based_order_book :: leftovers

### PR DESCRIPTION
**Pull request by GFariasR**
The following changes were left on my branch before the last pull request was accepted.

I think it deserves to take a look to see if we want them integrated. Let me know your thoughts.

Move config::validate to a cpp file.
Fix minor usage of is_ascending on UT.
Improve UT comments.
Fix misleading exception messages.
Move ssteam out of array_based_order_book.hpp to *.cpp

### CODE REVIEW

File: jb/itch5/array_based_order_book.cpp [hunk 1]
```diff
@@ -10,11 +9,39 @@ namespace itch5 {
...
+  if ((max_size() <= 0) or (max_size() > defaults::Max_Size)) {
+    std::ostringstream os;
+    os << "max-size must be > 0 and <=" << defaults::Max_Size
+       << ", value=" << max_size();
+    throw jb::usage(os.str(), 1);
+  }
+}
+
+/// handle exception with tk_begin_top_, px, and qty
+void handle_exception(
```
Comment (coryan) [inline jb/itch5/array_based_order_book.cpp]: I think "raise_exception" is a better name, please consider it.

File: jb/itch5/array_based_order_book.hpp [hunk 2]
```diff
@@ -22,14 +21,16 @@ namespace itch5 {
 
 // forward declaration of
 void validate_add_order_params(int, price4_t px = price4_t(0));
+void handle_exception(std::string, std::size_t, price4_t, int);
+void handle_exception(
+    std::string, std::size_t, std::size_t, std::size_t, price4_t, int);
 
 namespace defaults {
```
Comment (coryan) [inline jb/itch5/array_based_order_book.hpp]: I think it is possible to move the whole namespace defaults to the .cpp file, right?

File: jb/itch5/array_based_order_book.hpp [hunk 3]
```diff
@@ -22,14 +21,16 @@ namespace itch5 {
...
 
 #ifndef JB_ARRAY_DEFAULTS_max_size
 #define JB_ARRAY_DEFAULTS_max_size 10000
 #endif
-
-std::size_t max_size = JB_ARRAY_DEFAULTS_max_size;
+static std::size_t Max_Size = JB_ARRAY_DEFAULTS_max_size;
```
Comment (coryan) [inline jb/itch5/array_based_order_book.hpp]: Please do not Camel_Case names.  Let's use max_size for that variable.  Also, it should be std::size_t const, not static.

**Review by coryan — CHANGES_REQUESTED** [review]
Thanks!  That explains some stuff I was noticing.  A few minor changes requested.

File: jb/itch5/array_based_order_book.cpp (hunk 1 above)
Comment (GFariasR) [reply jb/itch5/array_based_order_book.cpp]: ACK.

Done.

File: jb/itch5/array_based_order_book.hpp (hunk 2 above)
Comment (GFariasR) [reply jb/itch5/array_based_order_book.hpp]: ACK.

Done

File: jb/itch5/array_based_order_book.hpp (hunk 3 above)
Comment (GFariasR) [reply jb/itch5/array_based_order_book.hpp]: ACK.

Done.

Comment (GFariasR) [reply jb/itch5/array_based_order_book.hpp]: BTW< what is this?

`  config_object_constructors(config);
`

Comment (GFariasR) [reply jb/itch5/array_based_order_book.hpp]: > Thanks! That explains some stuff I was noticing. A few minor changes requested.

What did you notice?

File: jb/itch5/array_based_order_book.cpp [hunk 4]
```diff
@@ -1,20 +1,65 @@
-#include <jb/itch5/price_field.hpp>
-#include <jb/itch5/quote_defaults.hpp>
-#include <jb/feed_error.hpp>
+#include <jb/itch5/array_based_order_book.hpp>
```
Comment (coryan) [inline jb/itch5/array_based_order_book.cpp]: Sorry, I missed this in the last review.  Use "" for this #include... That way clang-format knows that it should be the first one.

File: jb/itch5/array_based_order_book.hpp (hunk 3 above)
Comment (coryan) [reply jb/itch5/array_based_order_book.hpp]: > BTW< what is this?

A macro that defines the constructors for a config object.  They are always the same but have to be explicitly defined, I know it is yucky.

> What did you notice?

I thought I was missing some of the changes I had requested in the previous review.

**Review by coryan — CHANGES_REQUESTED** [review]
One other change, sorry I should have noticed in the previous pass.

**Review by coryan — APPROVED** [review]

**Pull request merged**